pkg/reconciler/pipeline/validate.go

```diff
@@ -107,6 +107,10 @@ func ValidatePipeline(pl *dfv1.Pipeline) error {
 				return fmt.Errorf("invalid user-defined source vertex %q, only one of 'http', 'kafka', 'nats', 'redisStreams', 'generator' and 'udSource' can be specified", k)
 			}
 		}
+
+		if findForests(s, pl) {
```

Comment:
Do we need to `findForests` for all the sources? Can we define the findForests as below?

```
func findForests(pl *dfv1.Pipeline) bool
```

Inside the function, we randomly pick a vertex and DFS. If the pipeline is a forest, then ANY of the vertices cannot visit ALL vertices. Hence if the eventual lengths don't match, the pipeline is a forest.

Reply:
Correct. I didn't notice we were doing it for all sources. We should only do it for one vertex.

Reply:
In this case, how would we randomly pick the vertex? I'm in favor of just calling it on `pl.Spec.Vertices[0]` each time as well.

Reply:
that's fine